src/Polly.Core/ResiliencePipelineBuilderBase.cs

```diff
@@ -129,13 +128,15 @@ internal PipelineComponent BuildPipelineComponent()
 
         var source = new ResilienceTelemetrySource(Name, InstanceName, null);
 
-        return PipelineComponentFactory.CreateComposite(components, new ResilienceStrategyTelemetry(source, TelemetryListener), TimeProvider);
+        return PipelineComponentFactory.CreateComposite(components, new ResilienceStrategyTelemetry(source, TelemetryListener), GetTimeProvider());
     }
 
+    internal TimeProvider GetTimeProvider() => TimeProvider ?? TimeProvider.System;
```

Comment:
What about `Resolve` instead of `Get`? Or just having a property called `TimeProviderInternal` or something?